python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/asset_daemon_scenario.py

```diff
@@ -0,0 +1,246 @@
+            for arr, err in zip(sorted_run_requests, sorted_expected_run_requests):
+                assert set(arr.asset_selection or []) == set(err.asset_selection or [])
+                assert arr.partition_key == err.partition_key
+        except:
+            self._log_assertion_error(sorted_expected_run_requests, sorted_run_requests)
+            raise
+
+        return self
+
+    def assert_evaluation(
+        self,
+        key: CoercibleToAssetKey,
+        expected_evaluation_specs: Sequence[AssetRuleEvaluationSpec],
+        num_requested: Optional[int] = None,
+        num_skipped: Optional[int] = None,
+        num_discarded: Optional[int] = None,
+    ) -> "AssetDaemonScenarioState":
+        """Makes an assertion about the evaluation data associated with a specific asset."""
+        asset_key = AssetKey.from_coercible(key)
+        actual_evaluation = next((e for e in self.evaluations if e.asset_key == asset_key), None)
+        if actual_evaluation is None:
+            try:
+                assert len(expected_evaluation_specs) == 0
+                assert all(n is None for n in (num_requested, num_skipped, num_discarded))
+            except:
+                self.logger.error(
+                    "\nAll Evaluations: \n\n" + "\n\n".join("\t" + str(e) for e in self.evaluations)
+                )
+                raise
+            return self
+        if num_requested is not None:
+            assert actual_evaluation.num_requested == num_requested
+        if num_skipped is not None:
+            assert actual_evaluation.num_skipped == num_skipped
+        if num_discarded is not None:
+            assert actual_evaluation.num_discarded == num_discarded
+
+        # unpack the serialized partition subsets into an easier format
+        actual_rule_evaluations = [
+            (
+                rule_evaluation,
+                sorted(
+                    serialized_subset.deserialize(
+                        check.not_none(self.asset_graph.get_partitions_def(asset_key))
+                    ).get_partition_keys()
+                )
+                if serialized_subset is not None
+                else None,
+            )
+            for rule_evaluation, serialized_subset in actual_evaluation.partition_subsets_by_condition
+        ]
+        expected_rule_evaluations = [ees.resolve() for ees in expected_evaluation_specs]
+
+        try:
+            for (ared, aps), (ered, eps) in zip(
+                sorted(actual_rule_evaluations), sorted(expected_rule_evaluations)
+            ):
+                assert ared.rule_snapshot == ered.rule_snapshot
+                assert aps == eps
+                # only check evaluation data if it was set on the expected evaluation spec
+                if ered.evaluation_data is not None:
+                    assert ared.evaluation_data == ered.evaluation_data
+
+        except:
+            self._log_assertion_error(
```

Comment:
this will just print strings out containing the expected rules and actual rules, right?

This could potentially still be a lot of text... if it's easy to add a way to show the different/missing specific rules, that would be very nice

Reply:
this is non-blocking

Reply:
Definitely would be a nice upgrade / fairly easy -- will leave that for a later PR / whenever one of us gets annoyed enough to implement it.